primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -652,6 +651,344 @@ where
     }
 }
 
+#[cfg(feature = "icicle")]
+pub(crate) mod icicle {
+    use super::*;
+    use crate::icicle_deps::{curves::*, *};
+    use ark_ec::{short_weierstrass::Affine, CurveConfig};
+    use ark_ff::BigInteger;
+
+    /// Trait for GPU-accelerated PCS.commit APIs
+    pub trait GPUCommit<E, C>
+    where
+        C: IcicleCurve + MSM<C>,
+        C::ScalarField: ArkConvertible<ArkEquivalent = E::ScalarField>,
+        C::BaseField: ArkConvertible<ArkEquivalent = <C::ArkSWConfig as CurveConfig>::BaseField>,
+        <C::ArkSWConfig as CurveConfig>::BaseField: PrimeField,
+        E: Pairing<G1Affine = Affine<<C as IcicleCurve>::ArkSWConfig>>,
+    {
+        /// The full cycle of computing poly-commit on GPU
+        fn commit_with_gpu(
+            prover_param: impl Borrow<UnivariateProverParam<E>>,
+            poly: &DensePolynomial<E::ScalarField>,
+        ) -> Result<Commitment<E>, PCSError> {
+            let stream = warmup_new_stream().unwrap();
+
+            #[cfg(feature = "kzg-print-trace")]
+            let commit_time =
+                start_timer!(|| format!("Committing to polynomial of degree {} ", poly.degree()));
+
+            let mut srs_on_gpu = Self::load_prover_param_to_gpu(prover_param, poly.degree())?;
+            let poly_on_gpu = Self::load_poly_to_gpu(poly)?;
+            let msm_result_on_gpu = Self::commit_on_gpu(&mut srs_on_gpu, &poly_on_gpu, 1, &stream)?;
+            let comm = Self::load_commitments_to_host(msm_result_on_gpu, &stream)?[0];
+
+            #[cfg(feature = "kzg-print-trace")]
+            end_timer!(commit_time);
+
+            Ok(comm)
+        }
+
+        /// Similar to [`Self::commit_with_gpu()`] but for a batch of poly of
+        /// the same degree
+        fn batch_commit_with_gpu(
+            prover_param: impl Borrow<UnivariateProverParam<E>>,
+            polys: &[DensePolynomial<E::ScalarField>],
+        ) -> Result<Vec<Commitment<E>>, PCSError> {
+            let stream = warmup_new_stream().unwrap();
+
+            let degree = polys[0].degree();
```

Comment:
Maybe add a check that all polys are having the same degree?

Reply:
ah, this check was already in the `load_batch_poly_to_gpu()`, do we want duplicated check here as well? (even tho it should be blazing fast, negl cost) 
I'm fine either way

Reply:
check added in [ae6a9ca](https://github.com/EspressoSystems/jellyfish/pull/498/commits/ae6a9cadc4fa54aef719af0cce53e5a8b075bbfb)